File: service/collection_backup.py
"""Single authorized device, immutable full collection revisions in the shared SQLite DB."""
import json
import math
import time
import uuid


def prepare(db):
    db.execute('CREATE TABLE IF NOT EXISTS collection_snapshots (revision INTEGER PRIMARY KEY, device_id TEXT NOT NULL, snapshot TEXT NOT NULL, received_at REAL NOT NULL)')


def validate(snapshot):
    if set(snapshot) != {'deviceID', 'revision', 'favorites'}:
        raise ValueError('invalid snapshot')
    if str(uuid.UUID(snapshot['deviceID'])) != snapshot['deviceID'] or type(snapshot['revision']) is not int or not 0 < snapshot['revision'] < 2**63-1:
        raise ValueError('invalid identity or revision')
    if not isinstance(snapshot['favorites'], list):
        raise ValueError('invalid favorites')
    ids = set()
    for item in snapshot['favorites']:
        poem = item['poem']
        if set(item) != {'poem', 'savedAt'} or type(item['savedAt']) not in (int, float) or not math.isfinite(item['savedAt']):
            raise ValueError('invalid saved time')
        if not all(isinstance(poem[k], str) and poem[k] for k in ('id', 'title', 'poetID', 'poet', 'dynasty')) or poem['id'] in ids:
            raise ValueError('invalid poem')
        if not isinstance(poem['lines'], list) or not poem['lines'] or not all(isinstance(s, str) and s for s in poem['lines']):
            raise ValueError('invalid poem lines')
        if not isinstance(poem['sources'], list) or not all(all(isinstance(s[k], str) for k in ('title', 'url', 'license', 'note')) for s in poem['sources']):
            raise ValueError('invalid sources')
        if poem.get('background') is not None and not isinstance(poem['background'], str):
            raise ValueError('invalid background')
        ids.add(poem['id'])


def latest(db):
    prepare(db)
    row = db.execute('SELECT snapshot, received_at FROM collection_snapshots ORDER BY revision DESC LIMIT 1').fetchone()
    return None if row is None else dict(snapshot=json.loads(row[0]), receivedAt=row[1])


def save(db, snapshot):
    validate(snapshot)
    prepare(db)
    db.execute('BEGIN IMMEDIATE')
    current = latest(db)
    if current:
        previous = current['snapshot']
        if previous == snapshot:
            return current
        if previous['deviceID'] != snapshot['deviceID'] or snapshot['revision'] <= previous['revision']:
            raise ValueError('device or revision conflict')
    receipt = dict(snapshot=snapshot, receivedAt=time.time())
    db.execute('INSERT INTO collection_snapshots VALUES (?,?,?,?)', (snapshot['revision'], snapshot['deviceID'], json.dumps(snapshot, allow_nan=False), receipt['receivedAt']))
    return receipt

File: Sources/PoetryCore/CollectionBackup.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct CollectionSnapshot: Codable, Equatable {
    public let deviceID: String
    public let revision: Int64
    /// Newest first, including complete offline poem snapshots.
    public let favorites: [FavoritePoem]
}
public struct CollectionBackupReceipt: Codable, Equatable {
    public let snapshot: CollectionSnapshot
    public let receivedAt: Double
}
public struct CollectionBackupStatus {
    public let pending: Bool
    public let lastSuccess: Date?
}

extension PoetryApplication {
    func prepareBackup() throws {
        try prepareCollection()
        try database.execute("CREATE TABLE IF NOT EXISTS collection_backup (singleton INTEGER PRIMARY KEY CHECK(singleton=1), device TEXT NOT NULL, revision INTEGER NOT NULL, acknowledged INTEGER NOT NULL, success REAL)")
        try database.execute("INSERT OR IGNORE INTO collection_backup VALUES (1, ?, CASE WHEN EXISTS(SELECT 1 FROM favorites) THEN 1 ELSE 0 END, 0, NULL)", [.text(UUID().uuidString.lowercased())])
    }
    func advanceCollectionRevision() throws {
        try database.execute("UPDATE collection_backup SET revision = revision + 1 WHERE singleton=1")
    }
    public func collectionSnapshot() throws -> CollectionSnapshot {
        try prepareBackup()
        let row = try database.query("SELECT * FROM collection_backup")[0]
        return CollectionSnapshot(deviceID: row["device"]!.string!, revision: row["revision"]!.integer!, favorites: try favoritePoems())
    }
    public func collectionBackupStatus() throws -> CollectionBackupStatus {
        try prepareBackup()
        let row = try database.query("SELECT * FROM collection_backup")[0]
        let time: Date?
        if case .real(let value) = row["success"] { time = Date(timeIntervalSince1970: value) } else { time = nil }
        return CollectionBackupStatus(pending: row["revision"]!.integer! > row["acknowledged"]!.integer!, lastSuccess: time)
    }
    public func acknowledgeCollectionBackup(_ receipt: CollectionBackupReceipt, sent: CollectionSnapshot) throws {
        guard receipt.snapshot == sent, receipt.receivedAt.isFinite else { throw ContentUpdateError.invalid }
        try prepareBackup()
        try database.execute("UPDATE collection_backup SET acknowledged=MAX(acknowledged, ?), success=? WHERE device=? AND revision>=?", [.integer(sent.revision), .real(receipt.receivedAt), .text(sent.deviceID), .integer(sent.revision)])
    }
    /// Explicit user operation. The canonical service identity survives reinstall through this receipt.
    public func restoreCollection(_ receipt: CollectionBackupReceipt) throws {
        let snapshot = receipt.snapshot
        guard UUID(uuidString: snapshot.deviceID) != nil, snapshot.revision > 0, receipt.receivedAt.isFinite,
              Set(snapshot.favorites.map(\.id)).count == snapshot.favorites.count,
              snapshot.favorites.allSatisfy({ !$0.id.isEmpty && !$0.poem.title.isEmpty && !$0.poem.lines.isEmpty && $0.savedAt.timeIntervalSince1970.isFinite }) else { throw ContentUpdateError.invalid }
        _ = try favoritePoems()
        try prepareBackup()
        try database.transaction {
            try database.execute("DELETE FROM favorites")
            for item in snapshot.favorites.reversed() {
                let payload = String(decoding: try JSONEncoder().encode(item.poem), as: UTF8.self)
                try database.execute("INSERT INTO favorites (poem_id,payload,saved_at) VALUES (?,?,?)", [.text(item.id), .text(payload), .real(item.savedAt.timeIntervalSince1970)])
            }
            try database.execute("UPDATE collection_backup SET device=?,revision=?,acknowledged=?,success=?", [.text(snapshot.deviceID), .integer(snapshot.revision), .integer(snapshot.revision), .real(receipt.receivedAt)])
        }
    }
}

public final class CollectionBackupClient {
    private let credentials: ServiceCredentials
    private let session: URLSession
    public init(credentials: ServiceCredentials, session: URLSession? = nil) {
        self.credentials = credentials
        self.session = session ?? URLSession(configuration: .ephemeral, delegate: BackupRedirectGuard(), delegateQueue: nil)
    }
    public func fetch() async throws -> CollectionBackupReceipt { try await request(snapshot: nil) }
    public func upload(_ snapshot: CollectionSnapshot) async throws -> CollectionBackupReceipt { try await request(snapshot: snapshot) }
    private func request(snapshot: CollectionSnapshot?) async throws -> CollectionBackupReceipt {
        var request = URLRequest(url: credentials.baseURL.appendingPathComponent("v1/backup"))
        request.setValue("Bearer " + credentials.token, forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 30
        if let snapshot {
            request.httpMethod = "PUT"
            request.httpBody = try JSONEncoder().encode(snapshot)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: request)
        guard let response = response as? HTTPURLResponse, response.statusCode == 200, data.count <= 8_000_000 else { throw ContentUpdateError.unavailable }
        return try JSONDecoder().decode(CollectionBackupReceipt.self, from: data)
    }
}

private final class BackupRedirectGuard: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) { completionHandler(nil) }
}
